### ISSUE TITLE
Garbled text / buffering issue with fast updates and colored text

### ISSUE THREAD

**Issue by jimhester** [rstudio/rstudio#4777]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.12.1335
    OS Version      :  macOS Mojave
    R Version       : 3.6

### Steps to reproduce the problem

Run the following code in the RStudio console pane. The first two `show_progress()` calls should work as expected, the third generates garbled text.

```r
library(crayon)
library(progress)

show_progress <- function(format, size, width = 80) {
  pb <- progress_bar$new(total = size, format = format, width = width)
  for (i in 1:size) {
    pb$tick()
    Sys.sleep(1 / size)
  }
}

# Works with colors and 1 ms updates
show_progress(":bar " %+% green(":rate"), 1000)

# Works without colors and .1 ms updates
show_progress(":bar " %+% ":rate", 10000)

# Doesn't work with colors and .1 ms updates
show_progress(":bar " %+% green(":rate"), 10000)
```

The same code works fine in a terminal outside of RStudio and even in a R session started in the RStudio terminal pane.

My guess is there is some sort of buffering going on in the RStudio console that is messing up the updates when they happen quickly enough.

**Comment by gtritchie** [comment]
This is a bug in the IDE's handling of carriage returns with a line containing different colors, and is triggered by the progress-bar output being inconsistent in length (the part before the colored text):

For example, here's a subset of what is being written to the console in the third scenario, notice the difference in output length; every-time this happens the console messes up and doesn't overwrite the previous output.

```
"====>----------------------------------------------------------------- "
"====>------------------------------------------------------------------ "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
"====>-------------------------------------------------------------------- "
"====>----------------------------------------------------------------- "
```

**Comment by gtritchie** [comment]
Also note that all of this (potentially massive) amount of output is being captured as a single line of ever-growing text by the console  buffer, and won't reply correctly when it exceed our internal line-length; that is, if you refresh the browser during or after one of these progress bars having been used.

**Comment by gtritchie** [comment]
Fixed in 1.2-patch via 0be83fa.

**Comment by ronblum** [comment]
Verified in RStudio Desktop 1.2.1526 on Windows 10.
